prometheus/prometheus.go

```diff
@@ -0,0 +1,289 @@
+	defer db.Close()
+	if err != nil {
+		log.WithError(err).Error("Failed to connect to db")
+		return
+	}
+	messageCount.WithLabelValues(server, channel).Inc()
+	_, err = db.Exec("INSERT INTO messageCount VALUES(" + server + ", " + channel + ", 1) ON DUPLICATE KEY UPDATE value = value + 1;")
+	if err != nil {
+		log.WithError(err).Error("Failed to update messageCount")
+		return
+	}
+}
+
+// MessageDelete is called whenever a message is deleted
+// Decrements messageCount for the given server and channel
+func MessageDelete(server string, channel string) {
+	db, err := sql.Open("mysql", fmt.Sprintf("%s:%s@%s/%s", viper.GetString("mysql.username"), viper.GetString("mysql.password"), viper.GetString("mysql.url"), viper.GetString("prom.dbname")))
+	defer db.Close()
+	if err != nil {
+		log.WithError(err).Error("Failed to connect to db")
+		return
+	}
+
+	messageCount.WithLabelValues(server, channel).Dec()
+	_, err = db.Exec("INSERT INTO messageCount VALUES(" + server + ", " + channel + ", 0) ON DUPLICATE KEY UPDATE value = value - 1;")
+	if err != nil {
+		log.WithError(err).Error("Failed to update messageCount")
+		return
+	}
+}
+
+func createTables() {
+	db, err := sql.Open("mysql", fmt.Sprintf("%s:%s@%s/%s", viper.GetString("mysql.username"), viper.GetString("mysql.password"), viper.GetString("mysql.url"), viper.GetString("prom.dbname")))
+	defer db.Close()
+	if err != nil {
+		log.WithError(err).Error("Failed to connect to db")
+		return
+	}
+	_, err = db.Exec("CREATE TABLE IF NOT EXISTS stats(name VARCHAR(20) PRIMARY KEY, value INT);")
+	if err != nil {
+		log.WithError(err).Error("Failed to connect to create table stats")
+		return
+	}
+	_, err = db.Exec("CREATE TABLE IF NOT EXISTS messageCount(server VARCHAR(20), channel VARCHAR(20), value INT, PRIMARY KEY (server, channel));")
+	if err != nil {
+		log.WithError(err).Error("Failed to create table messageCount")
+		return
+	}
+	_, err = db.Exec("CREATE TABLE IF NOT EXISTS joined(id VARCHAR(20));")
+	if err != nil {
+		log.WithError(err).Error("Failed to create table joined")
+		return
+	}
+}
+
+func setup(s *discordgo.Session) {
+	globalSession = s
+	createTables()
+	db, err := sql.Open("mysql", fmt.Sprintf("%s:%s@%s/%s", viper.GetString("mysql.username"), viper.GetString("mysql.password"), viper.GetString("mysql.url"), viper.GetString("prom.dbname")))
+	defer db.Close()
+	if err != nil {
+		log.WithError(err).Error("Failed to connect to db")
+		return
+	}
+	servers := viper.Get("discord.servers").(*config.Servers)
+	publicServer, err := s.Guild(servers.PublicServer)
+	if err != nil {
+		log.WithError(err).Error("Failed to get Public Server guild")
+		return
+	}
+	for _, member := range publicServer.Members {
+		found := false
+		for _, roleID := range strings.Split(viper.GetString("discord.roles"), ",") {
+			for _, role := range member.Roles {
+				if role == roleID {
+					found = true
+					break
+				}
+			}
+			if found {
+				break
+			}
+		}
+		if found {
+			MemberJoin(member.User.ID)
```

Comment:
This will create a new DB connection for each member in the public server

Reply:
Actually on this one - wouldn't it make sense to do a bulk insert rather than a separate query for each one?

Reply:
Yea it probably makes sense to do a bulk insert, although I would kind of have to copy a lot of the MemberJoin function to this part, not sure it makes much difference in terms of performance

Reply:
Won't it run ~300 queries the first time you run the setup function for all the members? Not sure that would be ideal when it could be done as a single query.